### PR TITLE
[Profiler] Improve .NET Framework profiling support

### PR DESCRIPTION
**Pull request by chrisnas**
## Summary of changes
- Fix empty stack issue for lock contention samples
- Support garbage collection events
- Support allocation events

## Reason for change
Improve .NET Framework profiling support 
- PROF-8652
- PROF-6983

### CODE REVIEW

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/ContentionProvider.cpp [hunk 1]
```diff
@@ -156,8 +156,11 @@ void ContentionProvider::AddContentionSample(uint64_t timestamp, uint32_t thread
         // We know that we don't have any span ID nor end point details
 
         rawSample.Timestamp = timestamp;
-        auto end_stack = stack.begin() + std::min(stack.size(), static_cast<std::size_t>(rawSample.Stack.Capacity()));
-        std::copy(stack.begin(), end_stack, rawSample.Stack.begin());
+        auto cs = _callstackProvider.Get();
+        auto end_stack = stack.begin() + std::min(stack.size(), static_cast<std::size_t>(cs.Capacity()));
+        std::copy(stack.begin(), end_stack, cs.begin());
+        cs.SetCount(std::min(stack.size(), static_cast<std::size_t>(cs.Capacity())));
+        rawSample.Stack = std::move(cs);
```
Comment (gleocadie) [inline profiler/src/ProfilerEngine/Datadog.Profiler.Native/ContentionProvider.cpp:163]: Maybe we can put the size/nb frames in a variable instead
```suggestion
        const auto nbFrames = std::min(stack.size(), static_cast<std::size_t>(cs.Capacity());
        auto end_stack = stack.begin() + nbFrames);
        std::copy(stack.begin(), end_stack, cs.begin());
        cs.SetCount(nbFrames);
        rawSample.Stack = std::move(cs);
```

**Review by gleocadie — APPROVED** [review]
What are the changes related to GC ? (I read it in the description)

**Comment by andrewlock** [comment]
## Execution-Time Benchmarks Report :stopwatch:

Execution-time results for samples comparing the following branches/commits:
- [This PR (5867)](https://github.com/DataDog/dd-trace-dotnet/tree/be5162e2046a624d73188173cba7bced8ea5c875)
- [master](https://github.com/DataDog/dd-trace-dotnet/tree/0261bd2d097f759ea32153c262f045a2d1d5f714)

Execution-time benchmarks measure the whole time it takes to execute a program. And are intended to measure the one-off costs. Cases where the execution time results for the PR are worse than latest master results are shown in **red**. The following thresholds were used for comparing the execution times:
* Welch test with statistical test for significance of **5%**
* Only results indicating a difference greater than **5%** and **5 ms** are considered.


Note that these results are based on a _single_ point-in-time result for each branch. For full results, see the [dashboard](https://ddstaging.datadoghq.com/dashboard/4qn-6fi-54p/apm-net-execution-time-benchmarks).

Graphs show the p99 interval based on the mean and StdDev of the test run, as well as the mean value of the run (shown as a diamond below the graph).
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET Framework 4.6.2) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5867) - mean (71ms)  : 64, 79
     .   : milestone, 71,
    master - mean (72ms)  : 63, 81
     .   : milestone, 72,

    section CallTarget+Inlining+NGEN
    This PR (5867) - mean (1,067ms)  : 1041, 1093
     .   : milestone, 1067,
    master - mean (1,067ms)  : 1046, 1087
     .   : milestone, 1067,

```
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET Core 3.1) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5867) - mean (110ms)  : 104, 116
     .   : milestone, 110,
    master - mean (110ms)  : 105, 115
     .   : milestone, 110,

    section CallTarget+Inlining+NGEN
    This PR (5867) - mean (760ms)  : 732, 788
     .   : milestone, 760,
    master - mean (756ms)  : 727, 786
     .   : milestone, 756,

```
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET 6) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5867) - mean (92ms)  : 89, 95
     .   : milestone, 92,
    master - mean (93ms)  : 89, 96
     .   : milestone, 93,

    section CallTarget+Inlining+NGEN
    This PR (5867) - mean (701ms)  : 685, 717
     .   : milestone, 701,
    master - mean (704ms)  : 681, 728
     .   : milestone, 704,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET Framework 4.6.2) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5867) - mean (193ms)  : 189, 198
     .   : milestone, 193,
    master - mean (192ms)  : 189, 195
     .   : milestone, 192,

    section CallTarget+Inlining+NGEN
    This PR (5867) - mean (1,173ms)  : 1141, 1204
     .   : milestone, 1173,
    master - mean (1,167ms)  : 1141, 1193
     .   : milestone, 1167,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET Core 3.1) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5867) - mean (278ms)  : 273, 283
     .   : milestone, 278,
    master - mean (276ms)  : 271, 281
     .   : milestone, 276,

    section CallTarget+Inlining+NGEN
    This PR (5867) - mean (918ms)  : 899, 938
     .   : milestone, 918,
    master - mean (922ms)  : 903, 941
     .   : milestone, 922,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET 6) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (5867) - mean (268ms)  : 260, 275
     .   : milestone, 268,
    master - mean (266ms)  : 262, 270
     .   : milestone, 266,

    section CallTarget+Inlining+NGEN
    This PR (5867) - mean (912ms)  : 882, 943
     .   : milestone, 912,
    master - mean (908ms)  : 886, 929
     .   : milestone, 908,

```

**Comment by andrewlock** [comment]
## Throughput/Crank Report :zap:

Throughput results for AspNetCoreSimpleController comparing the following branches/commits:
- [This PR (5867)](https://github.com/DataDog/dd-trace-dotnet/tree/be5162e2046a624d73188173cba7bced8ea5c875)
- [master](https://github.com/DataDog/dd-trace-dotnet/tree/0261bd2d097f759ea32153c262f045a2d1d5f714)
- [benchmarks/2.9.0](https://github.com/DataDog/dd-trace-dotnet/tree/89ae3f4ee7d85b824badda5873166707c9e560b4)

Cases where throughput results for the PR are worse than latest master (5% drop or greater), results are shown in **red**.

Note that these results are based on a _single_ point-in-time result for each branch. For full results, see [one](https://ddstaging.datadoghq.com/dashboard/fnz-afi-c2i/apm-net-crank) of the [many](https://ddstaging.datadoghq.com/dashboard/c5g-i7d-kcy/ciapp-apm-net-throughput-tests), [many](https://ddstaging.datadoghq.com/dashboard/uxh-m8j-qhi/ciapp-apm-net-throughput-tests-kevin) dashboards!

```mermaid
gantt
    title Throughput Linux x64 (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (5867) (11.558M)   : 0, 11557672
    master (11.989M)   : 0, 11988922
    benchmarks/2.9.0 (11.513M)   : 0, 11512780

    section Automatic
    This PR (5867) (7.709M)   : crit ,0, 7708941
    master (8.129M)   : 0, 8128534
    benchmarks/2.9.0 (8.207M)   : 0, 8206882

    section Trace stats
    master (8.363M)   : 0, 8362615

    section Manual
    master (12.022M)   : 0, 12022316

    section Manual + Automatic
    This PR (5867) (7.155M)   : 0, 7155247
    master (7.524M)   : 0, 7524228

    section DD_TRACE_ENABLED=0
    master (11.062M)   : 0, 11062172

```
```mermaid
gantt
    title Throughput Linux arm64 (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (5867) (9.466M)   : 0, 9466191
    benchmarks/2.9.0 (9.495M)   : 0, 9494663

    section Automatic
    This PR (5867) (6.637M)   : 0, 6636570

    section Manual + Automatic
    This PR (5867) (6.189M)   : 0, 6188918

```
```mermaid
gantt
    title Throughput Windows x64 (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (5867) (10.352M)   : 0, 10352445

    section Automatic
    This PR (5867) (5.988M)   : 0, 5988210

    section Manual + Automatic
    This PR (5867) (6.491M)   : 0, 6490788

```

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native.Windows/EtwEventsManager.cpp [hunk 2]
```diff
@@ -63,6 +63,7 @@ uint64_t TimestampToEpochNS(uint64_t eventTimestamp)
     t.tm_hour = st.wHour;
     t.tm_min = st.wMinute;
     t.tm_sec = st.wSecond;
+    t.tm_isdst = -1; // don't mess with daylight saving time (already done by SystemTimeToTzSpecificLocalTime)
```
Comment (r1viollet) [inline profiler/src/ProfilerEngine/Datadog.Profiler.Native.Windows/EtwEventsManager.cpp:66]: Nice! we figured it out!

**Comment by chrisnas** [comment]
> What are the changes related to GC ? (I read it in the description)

the timestamp fix works for both (gc events were recorded by with the wrong timestamp did not appear in the timeline)

**Pull request merged**